### PR TITLE
Add support for NAT traversal techniques

### PR DESCRIPTION
**Pull request by Rucadi**
**What is the purpose of this pull request? (put an "X" next to item)**

[X] New feature

**What changes did you make? (Give an overview)**
There is this PR: https://github.com/webtorrent/webtorrent/pull/1437 which added the feature of UPnP and NAT-PMP on webtorrent (node). 
I upgraded it a little since some names used changed and reopened the PR.

**Which issue (if any) does this pull request address?**

Adds UPnP to webtorrent.

### CODE REVIEW

**Comment by Rucadi** [comment]
I'm not too used to programming with JS, I saw some problems running the npm test and I fixed them, also checked that this is not used on the browser (This should be the latest two commits, before I just checked that it worked on node and assumed the code written by oleiba was already valid for all cases)

**Comment by DiegoRBaquero** [comment]
How did you test this? Would be great to have some unit tests

**Comment by Rucadi** [comment]
I don't know which could be the best way to create some unit tests for this.

The way I tested it was to upgrade the code from the previous PR, run some torrents and to check my router configuration, which was being updated as various ports were used.

I needed this feature and rescued, but I'm not expert on this matter. 

If anyone has any suggestion, I could look into it.

File: lib/nat-traversal.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
+const isBrowser = typeof process === 'undefined' || process.type === 'renderer' || process.browser === true || process.__nwjs
+
+if (isBrowser) {
+  module.exports = null
+} else {
+  const arrayRemove = require('unordered-array-remove')
+  const debug = require('debug')('webtorrent:nat-traversal')
+  const natUpnp = require('nat-upnp') // browser exclude
+  const natpmp = require('nat-pmp') // browser exclude
+  const network = require('network') // browser exclude
```
Comment (mrbaguvix) [inline lib/nat-traversal.js:10]: this should really not be here, move them to the top

File: lib/nat-traversal.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    // Refresh the mapping 10 minutes before the end of its lifetime
+    self.timeout = (self.ttl - 600) * 1000
+    self._openedPorts = []
+    self._intervalsUpnp = {}
+    self._intervalsPmp = {}
+
+    self._upnpPortMapping = function (port, protocol, cb) {
+      const self = this
+
+      debug('Mapping port %d for protocol %s on router using UPnP', port, protocol)
```
Comment (mrbaguvix) [inline lib/nat-traversal.js:30]: what is the use of debug()?

File: lib/nat-traversal.js (hunk 1 above)
Comment (DiegoRBaquero) [reply lib/nat-traversal.js:10]: why? they are conditionally required

File: lib/nat-traversal.js (hunk 2 above)
Comment (DiegoRBaquero) [reply lib/nat-traversal.js:30]: logging

**Comment by stamepicmorg** [comment]
if this feature will be merget - it will be aresome

**Comment by DiegoRBaquero** [comment]
Thoughts on this? https://github.com/achingbrain/nat-port-mapper

**Comment by resession** [comment]
any progress?

**Comment by resession** [comment]
can this be added?

**Comment by stamepicmorg** [comment]
I quit my job where this library was used. and the pull request was not added here. sadly.

**Comment by resession** [comment]
this would make webtorrent much more functional and complete

**Comment by ThaUnknown** [comment]
I could probably fairly easily fix this PR, but there are 2 issues: no tests, and I have no clue what it accomplishes

**Comment by resession** [comment]
@Rucadi would you be able give us a description of what this accomplishes and make some tests. i assume this would help in connecting peer's who are not accessible/behind nat.

**Comment by resession** [comment]
> I could probably fairly easily fix this PR, but there are 2 issues: no tests, and I have no clue what it accomplishes

i assume it helps with connecting to peers that are behind a nat

**Comment by Rucadi** [comment]
> @Rucadi would you be able give us a description of what this accomplishes and make some tests. i assume this would help in connecting peer's who are not accessible/behind nat.

This allow peers to connect to peers behind a nat. This dynamically allocate a port which other peers can connect to stablish connection. 

Without this PR, you limit the number of peers seeding a torrent.

I would be grateful if you fix this PR and we can merge it :)

**Comment by resession** [comment]
@ThaUnknown is this something that can be fixed and merged?

**Comment by midzer** [comment]
Does UPnP have to be enabled on your NAT-Router for this PR?

Traversing NAT works for me by specifying STUN `iceServer` in your WebTorrent config.

**Comment by Rucadi** [comment]
> Does UPnP have to be enabled on your NAT-Router for this PR?
> 
> Traversing NAT works for me by specifying STUN `iceServer` in your WebTorrent config.

UPnP or NAT-PMP should be enabled on the router, there are no side effects if disabled.

**Comment by resession** [comment]
> Is this still relevant? If so, what is blocking it? Is there anything you can do to help move it forward?

it is still relevant

**Comment by ThaUnknown** [comment]
this was implemented in https://github.com/webtorrent/webtorrent/pull/2570

**Pull request closed without merging**